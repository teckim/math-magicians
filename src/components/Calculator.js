import { useState } from 'react';
import Stack from '@mui/material/Stack';
import Card from '@mui/material/Card';
import CardContent from '@mui/material/CardContent';
import Box from '@mui/material/Box';
import CalculatorButton from './CalculatorButton';
import CalculatorScreen from './CalculatorScreen';
import calculate from '../logic/calculate';

const BUTTONS = [
  ['AC', '+/-', '%', 'รท'],
  ['7', '8', '9', 'x'],
  ['4', '5', '6', '-'],
  ['1', '2', '3', '+'],
  ['0', '.', null, '='],
];

const Calculator = () => {
  const [state, setState] = useState({ total: 0, next: 0 });

  const handleInputChange = (event) => {
    setState({
      next: event.target.value,
    });
  };

  const handleClick = (operation) => {
    const calc = calculate(state, operation);

    setState(calc);
  };

  return (
    <Card sx={{ mx: 'auto', width: '300px' }}>
      <CardContent>
        <CalculatorScreen
          onChange={handleInputChange}
          value={state.next || state.total || 0}
          operation={state.operation}
        />
        <div>
          {BUTTONS.map((line, lineIndex) => (
            <Stack key={line} direction="row" justifyContent="space-between">
              {line.map((button, index) => {
                const isPrimary = index === 3 || button === '=' || lineIndex === 0;

                return (
                  <div key={button}>
                    {!button ? (
                      <Box width={48} sx={{ m: '8px' }} />
                    ) : (
                      <CalculatorButton
                        value={button}
                        isPrimary={isPrimary}
                        onClick={handleClick}
                        sx={{ ml: button === '=' && 'auto' }}
                      />
                    )}
                  </div>
                );
              })}
            </Stack>
          ))}
        </div>
      </CardContent>
    </Card>
  );
};

export default Calculator;
