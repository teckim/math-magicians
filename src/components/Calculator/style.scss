$row_height: 48px;
$primary-color: #f5913e;
$secondary-color: #e0e0e0;

.calculator {
  width: 360px;

  &__input {
    width: 100%;
    height: $row_height;
    text-align: right;
    color: #fff;

    &-wrapper {
      padding: 0 8px;
      background-color: #858693;
    }
  }

  &__actions {
    display: grid;
    grid-template-columns: repeat(4, 1fr);
  }

  &__btn {
    height: $row_height;
    display: flex;
    align-items: center;
    justify-content: center;
    font-size: 12px;
    background-color: $secondary-color;
    border: 1px solid #d1d1d2;

    &--primary {
      background-color: $primary-color;
    }

    &--span {
      grid-column: 1 / 3;
    }
  }
}
