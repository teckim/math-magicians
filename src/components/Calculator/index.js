import { useState } from 'react';
import PropTypes from 'prop-types';
import calculate from '../../logic/calculate';

import './style.scss';

const BUTTONS = ['AC', '+/-', '%', 'รท', '7', '8', '9', 'x', '4', '5', '6', '-', '1', '2', '3', '+', '0', '.', '='];

const Button = ({
  value, isPrimary, onClick,
}) => {
  const classNames = [
    'calculator__btn',
    isPrimary && 'calculator__btn--primary',
    value === '0' && 'calculator__btn--span',
  ]
    .filter((v) => !!v)
    .join(' ');

  return <button className={classNames} onClick={onClick} type="button" value={value}>{value}</button>;
};

Button.propTypes = {
  value: PropTypes.string.isRequired,
  isPrimary: PropTypes.bool,
  onClick: PropTypes.func,
};

Button.defaultProps = {
  isPrimary: false,
  onClick: () => {},
};

const Calculator = () => {
  const [state, setState] = useState({ total: 0, next: 0 });

  const handleInputChange = (event) => {
    setState({
      next: event.target.value,
    });
  };

  const handleClick = (event) => {
    const operation = event.target.value;
    const calc = calculate(state, operation);

    setState(calc);
  };

  return (
    <div className="calculator">
      <div className="calculator__input-wrapper">
        <input data-testid="calculator-input" className="calculator__input" name="equation" disabled onChange={handleInputChange} value={state.next || state.total || 0} />
      </div>
      <div className="calculator__actions">
        {
          BUTTONS.map((button, index) => {
            const isPrimary = (index + 1) % 4 === 0 || index === 18;

            return (
              <Button
                key={button}
                value={button}
                isPrimary={isPrimary}
                onClick={handleClick}
              />
            );
          })
        }
      </div>
    </div>
  );
};

export default Calculator;
