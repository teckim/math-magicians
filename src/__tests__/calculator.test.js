import React from 'react';
import TestRenderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';
import Calculator from '../components/Calculator';

describe('calculator component', () => {
  it('should render correctly', () => {
    const tree = TestRenderer.create(<Calculator />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('2 + 3 should be equale to 5', () => {
    render(<Calculator />);

    userEvent.click(screen.getByText('2'));
    userEvent.click(screen.getByText('+'));
    userEvent.click(screen.getByText('3'));
    userEvent.click(screen.getByText('='));

    const { value } = screen.getByTestId('calculator-input');

    expect(value).toBe('5');
  });

  test('5 - 3 should be equale to 2', () => {
    render(<Calculator />);

    userEvent.click(screen.getByText('AC'));
    userEvent.click(screen.getByText('5'));
    userEvent.click(screen.getByText('-'));
    userEvent.click(screen.getByText('3'));
    userEvent.click(screen.getByText('='));

    const { value } = screen.getByTestId('calculator-input');

    expect(value).toBe('2');
  });

  test('3 x 2 should be equale to 6', () => {
    render(<Calculator />);

    userEvent.click(screen.getByText('AC'));
    userEvent.click(screen.getByText('3'));
    userEvent.click(screen.getByText('x'));
    userEvent.click(screen.getByText('2'));
    userEvent.click(screen.getByText('='));

    const { value } = screen.getByTestId('calculator-input');

    expect(value).toBe('6');
  });

  test('8 รท 2 should be equale to 4', () => {
    render(<Calculator />);

    userEvent.click(screen.getByText('AC'));
    userEvent.click(screen.getByText('8'));
    userEvent.click(screen.getByText('รท'));
    userEvent.click(screen.getByText('2'));
    userEvent.click(screen.getByText('='));

    const { value } = screen.getByTestId('calculator-input');

    expect(value).toBe('4');
  });
});
