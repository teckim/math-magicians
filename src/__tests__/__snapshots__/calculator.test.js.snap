// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculator component should render correctly 1`] = `
<div
  className="calculator"
>
  <div
    className="calculator__input-wrapper"
  >
    <input
      className="calculator__input"
      data-testid="calculator-input"
      disabled={true}
      name="equation"
      onChange={[Function]}
      value={0}
    />
  </div>
  <div
    className="calculator__actions"
  >
    <button
      className="calculator__btn"
      onClick={[Function]}
      type="button"
      value="AC"
    >
      AC
    </button>
    <button
      className="calculator__btn"
      onClick={[Function]}
      type="button"
      value="+/-"
    >
      +/-
    </button>
    <button
      className="calculator__btn"
      onClick={[Function]}
      type="button"
      value="%"
    >
      %
    </button>
    <button
      className="calculator__btn calculator__btn--primary"
      onClick={[Function]}
      type="button"
      value="รท"
    >
      รท
    </button>
    <button
      className="calculator__btn"
      onClick={[Function]}
      type="button"
      value="7"
    >
      7
    </button>
    <button
      className="calculator__btn"
      onClick={[Function]}
      type="button"
      value="8"
    >
      8
    </button>
    <button
      className="calculator__btn"
      onClick={[Function]}
      type="button"
      value="9"
    >
      9
    </button>
    <button
      className="calculator__btn calculator__btn--primary"
      onClick={[Function]}
      type="button"
      value="x"
    >
      x
    </button>
    <button
      className="calculator__btn"
      onClick={[Function]}
      type="button"
      value="4"
    >
      4
    </button>
    <button
      className="calculator__btn"
      onClick={[Function]}
      type="button"
      value="5"
    >
      5
    </button>
    <button
      className="calculator__btn"
      onClick={[Function]}
      type="button"
      value="6"
    >
      6
    </button>
    <button
      className="calculator__btn calculator__btn--primary"
      onClick={[Function]}
      type="button"
      value="-"
    >
      -
    </button>
    <button
      className="calculator__btn"
      onClick={[Function]}
      type="button"
      value="1"
    >
      1
    </button>
    <button
      className="calculator__btn"
      onClick={[Function]}
      type="button"
      value="2"
    >
      2
    </button>
    <button
      className="calculator__btn"
      onClick={[Function]}
      type="button"
      value="3"
    >
      3
    </button>
    <button
      className="calculator__btn calculator__btn--primary"
      onClick={[Function]}
      type="button"
      value="+"
    >
      +
    </button>
    <button
      className="calculator__btn calculator__btn--span"
      onClick={[Function]}
      type="button"
      value="0"
    >
      0
    </button>
    <button
      className="calculator__btn"
      onClick={[Function]}
      type="button"
      value="."
    >
      .
    </button>
    <button
      className="calculator__btn calculator__btn--primary"
      onClick={[Function]}
      type="button"
      value="="
    >
      =
    </button>
  </div>
</div>
`;
